src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/CompilerTypeSystemContext.IntefaceThunks.cs

```diff
@@ -0,0 +1,287 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+using Internal.TypeSystem;
+using Internal.IL;
+using Internal.IL.Stubs;
+
+using Debug = System.Diagnostics.Debug;
+
+// Default interface method implementation thunks
+//
+// The problem with default interface methods and shared generic code is that for:
+//
+// interface IFoo<T>
+// {
+//     Type GetTheType() => typeof(T);
+// }
+//
+// The actual generated code when instantiated over a shareable instance (like object) is
+// just IFoo<__Canon>.GetTheType, for any shareable argument (there's no unique code
+// generated for IFoo<Object>/IFoo<string>/... - we just have IFoo<__Canon>).
+//
+// For the canonical code to know what the actual T is, we need to provide the instantiation
+// context somehow. We can't easily get it from `this` like we do for reference types
+// since the type might implement multiple instantiations of IFoo (`class Abc : IFoo<object>, IFoo<string> { }`)
+// and we wouldn't know which one we are executing for within the method body.
+//
+// So we end up passing the context same as for shared valuetype code (that also cannot
+// determine context from just `this`) - by adding an extra instantiation
+// argument. The actual code for IFoo<__Canon>.GetTheType looks something like this:
+//
+// Type IFoo__Canon__GetTheType(IFoo<__Canon> instance, EEType* context)
+// {
+//     return Type.GetTypeFromHandle(GetTypeHandleOfTInIFooCanon(context));
+// }
+//
+// Now we have a problem because this method expects an extra `context` argument
+// that will not be provided at the callsite, since the callsite doesn't know
+// where it will dispatch to (could be a non-default-interface-method).
+//
+// We solve this with an instantiating thunk. The instantiating thunk is the thing
+// we place in the vtable of the implementing type. The thunk looks like this:
+//
+// Type Abc_IFoo__Canon__GetTheType_Thunk(IFoo<__Canon> instance)
+// {
+//     return IFoo__Canon__GetTheType(instance, GetOrdinalInterface(instance.m_pEEType, 0));
+// }
+//
+// Notice the thunk now has the expected signature, and some code to compute the context.
+//
+// The GetOrdinalInterface method retrieves the specified interface EEType off the EEType's interface list.
+// The thunks are per-type (since the position in the inteface list is different).
+//
+// We hardcode the position in the interface list instead of just hardcoding the interface type
```

Comment:
If we were to use actual generic slots, we would need to make sure the injected method is reported from GetAllMethod on the type to make code like this work:

https://github.com/dotnet/runtimelab/blob/27148d32d6f7fbdf675989f4f34e70a1efa53120/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs#L407-L431

It's doable, but it's extra complexity since in GetAllMethods we don't quickly know whether there's any default-implemented methods. The above code is needed to ensure generic dictionaries look the same in multifile builds irrespective of what parts of the generic type were used.

Reply:
I agree that the way you have done it is nice and simple. I was just commenting on the part of the comment that says doing it the other way would require code generation.

Reply:
Yeah, I meant that if we were to record the exact instantiations in code - I didn't think about doing a generic lookup.